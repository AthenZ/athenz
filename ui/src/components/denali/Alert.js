/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import ReactModal from 'react-modal';
import Icon from '../denali/icons/Icon';
import {
    colors,
    cssDropShadow,
    cssFontStyles,
    cssFontWeights,
} from '../denali/styles';
import { css, cx } from 'emotion';
import { rgba } from 'polished';

const TYPE_COLOR_MAP = {
    danger: 'statusDanger',
    info: 'statusInfo',
    success: 'statusSuccess',
    warning: 'statusWarning',
};

const TYPE_ICON_MAP = {
    danger: 'stop-warning',
    info: 'information-circle',
    success: 'check-circle',
    warning: 'warning',
};

// http://reactcommunity.org/react-modal/examples/set_app_element.html
ReactModal.setAppElement('body');

const modalOverlayClass = css`
    background-color: transparent;
    bottom: 0;
    left: 0;
    position: fixed;
    right: 0;
    top: 0;
    z-index: 5;
    label: modal-overlay;
`;

const makeAlertClass = (props) => css`
    ${cssDropShadow};
    align-items: flex-start;
    background: ${colors.white};
    border-radius: 4px;
    border-top: 4px solid ${colors[TYPE_COLOR_MAP[props.type]]};
    box-sizing: border-box;
    color: ${colors.black};
    display: flex;
    flex-flow: row nowrap;
    ${cssFontStyles.default};
    left: 50%;
    max-height: calc(100vh - 160px);
    min-width: 400px;
    outline: none;
    position: fixed;
    transform: translate(-50%, 0);
    width: 400px;
    label: alert;
    &.ReactModal__Content {
        opacity: 0;
        top: 0;
    }
    &.animated {
        transition: opacity 0.2s ease-out,
            top 0.3s cubic-bezier(0.175, 0.885, 0.32, 1.275);
    }
    &.ReactModal__Content--after-open {
        top: 80px;
        opacity: 1;
    }
    /* Animate fade out on close. This is currently not used as tests fail */
    &.ReactModal__Content--before-close {
        opacity: 0;
    }
    & .left-section {
        flex: 0 0 28px;
        padding: 14px;
        & .alert-icon {
            align-items: flex-start;
            display: flex;
            justify-content: center;
        }
    }
    & .right-section {
        display: flex;
        flex: 1 1 auto;
        flex-flow: column nowrap;
        & .title-row {
            align-items: center;
            display: flex;
            flex-flow: row nowrap;
            height: initial;
            justify-content: space-between;
            padding: 14px 14px 14px 0;
            & .title {
                ${cssFontWeights.bold};
            }
        }
        & .description {
            font-size: 14px;
            font-weight: 300;
            padding: 0 14px 14px 0;
        }
    }
`;

/**
 * Alert displays a dismissable modal near the top of the screen. It is meant to
 * show a short blurb of information. There are four
 * alert types:
 *
 * 1. Info
 * 2. Success
 * 3. Warning
 * 4. Danger
 *
 * Under the hood, Alert uses [`react-modal`](http://reactcommunity.org/react-modal/).
 * To prevent the page from scrolling when the modal is showing, add the
 * following to the app's global CSS:
 *
 * ```css
 * .ReactModal__Body--open {
 *   overflow: hidden;
 * }
 */
class Alert extends React.Component {
    // Taken from:
    // https://stackoverflow.com/questions/45678517/better-way-to-cleartimeout-in-componentwillunmount
    setTimer = (ms) => {
        // Remember the timer handle
        this.timerHandle = setTimeout(() => {
            this.timerHandle = null;
            this.props.onClose();
        }, ms);
    };

    clearTimer = () => {
        // Is our timer running?
        if (this.timerHandle) {
            clearTimeout(this.timerHandle);
            this.timerHandle = null;
        }
    };

    componentDidMount() {
        if (this.props.isOpen && this.props.duration) {
            this.setTimer(this.props.duration);
        }
    }

    componentDidUpdate(prevProps) {
        if (!prevProps.isOpen && this.props.isOpen && this.props.duration) {
            this.setTimer(this.props.duration);
        } else if (prevProps.isOpen && !this.props.isOpen) {
            this.clearTimer();
        }
    }

    componentWillUnmount() {
        this.clearTimer();
    }

    render() {
        const StatusIcon = TYPE_ICON_MAP[this.props.type];
        const alertClass = makeAlertClass(this.props);

        return (
            <ReactModal
                aria={{
                    labelledby: 'alert-title',
                    describedby: 'alert-description',
                }}
                className={cx(
                    alertClass,
                    'denali-alert',
                    `denali-alert-${this.props.type}`,
                    {
                        animated: !this.props.noanim,
                    },
                    this.props.className
                )}
                closeTimeoutMS={200}
                descriptionLabel={`${this.props.type} alert`}
                isOpen={this.props.isOpen}
                overlayClassName={modalOverlayClass}
                shouldCloseOnOverlayClick={false}
                onRequestClose={this.props.onClose}
            >
                <div className='left-section'>
                    <div className='alert-icon' data-testid='alert-icon'>
                        <Icon
                            icon={'check-circle'}
                            color={colors[TYPE_COLOR_MAP[this.props.type]]}
                            decorative={true}
                            size='28px'
                            verticalAlign='middle'
                        />
                    </div>
                </div>
                <div className='right-section'>
                    <div className='title-row'>
                        <div
                            className='title'
                            id='alert-title'
                            data-testid='alert-title'
                        >
                            {this.props.title}
                        </div>
                        <div
                            className='close-button'
                            onClick={this.props.onClose}
                        >
                            <Icon
                                color={rgba(colors.grey800, 0.4)}
                                colorHover={rgba(colors.grey800, 0.6)}
                                isLink
                                size='28px'
                                icon={'close'}
                            />
                        </div>
                    </div>
                    {this.props.description && (
                        <div
                            className='description'
                            id='alert-description'
                            data-testid='alert-description'
                        >
                            {this.props.description}
                        </div>
                    )}
                </div>
            </ReactModal>
        );
    }
}

Alert.propTypes = {
    /** Additonal class to apply to the outer div of the modal content */
    className: PropTypes.string,
    /** Optional description. Can be a string or JSX / component */
    description: PropTypes.any,
    /** Once open, call `onClose()` after N milliseconds */
    duration: PropTypes.number,
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** Toggle the alert on or off */
    isOpen: PropTypes.bool,
    /** Title of the alert */
    title: PropTypes.string.isRequired,
    /** Alert supports four status types */
    type: PropTypes.oneOf(['danger', 'info', 'success', 'warning']),
    /** Handler for closing the alert */
    onClose: PropTypes.func.isRequired,
};

Alert.defaultProps = {
    duration: 0,
    isOpen: false,
    noanim: false,
    type: 'info',
};

export default Alert;
