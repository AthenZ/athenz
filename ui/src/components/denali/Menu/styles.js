/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { css } from 'emotion';
import { cssFontStyles } from '../styles/fonts';
import { colors } from '../styles/colors';
import { cssDropShadow } from '../styles/drop-shadow';

export const makeCssPopperBox = (props) => css`
    ${cssDropShadow};
    ${cssFontStyles.default};
    background-color: ${colors.white};
    border: 2px solid ${colors.border};
    border-radius: 4px;
    color: ${colors.black};
    cursor: default;
    line-height: 1.4;
    margin: 0;
    max-width: ${props.basic ? '400px' : undefined};
    min-width: 20px;
    padding: ${props.basic ? '6px 10px' : 0};
    transition: ${!props.noanim && 'opacity 0.25s ease-in'};
    z-index: 999;
    label: denali-menu;
    /* Shift the popper to allow space for arrow */
    &[data-placement*='bottom'] {
        margin-top: 8px;
    }
    &[data-placement*='top'] {
        margin-bottom: 8px;
    }
    &[data-placement*='left'] {
        margin-right: 8px;
    }
    &[data-placement*='right'] {
        margin-left: 8px;
    }

    /**
   * Invisible padding between trigger and main content (around the arrow) so
   * user can slide from one to the other without the menu closing.
   */
    &::after {
        background: transparent;
        content: '';
        position: absolute;
    }
    &[data-placement*='top']::after {
        bottom: -12px;
        height: 12px;
        width: 100%;
    }
    &[data-placement*='bottom']::after {
        height: 12px;
        top: -12px;
        width: 100%;
    }
    &[data-placement*='left']::after {
        right: -12px;
        height: 100%;
        width: 12px;
    }
    &[data-placement*='right']::after {
        left: -12px;
        height: 100%;
        width: 12px;
    }
`;

export const cssArrow = css`
    height: 1rem;
    position: absolute;
    width: 1rem;
    label: menu-arrow;

    &::before,
    &::after {
        border-style: solid;
        content: '';
        display: block;
        height: 0;
        margin: auto;
        position: absolute;
        width: 0;
    }

    &[data-placement*='bottom'] {
        left: 0;
        margin-top: -6px;
        top: -8px;
        width: 1rem;
    }

    &[data-placement*='bottom']::before {
        border-color: transparent transparent ${colors.grey400} transparent;
        border-width: 0 8px 8px 8px;
        position: absolute;
        top: 4px;
    }

    &[data-placement*='bottom']::after {
        border-color: transparent transparent ${colors.white} transparent;
        border-width: 0 8px 8px 8px;
        top: 7px;
    }

    &[data-placement*='top'] {
        bottom: 0;
        height: 1rem;
        left: 0;
        margin-bottom: -1rem;
        width: 1rem;
    }

    &[data-placement*='top']::before {
        border-color: ${colors.grey400} transparent transparent transparent;
        border-width: 8px 8px 0 8px;
        position: absolute;
        top: 2px;
    }

    &[data-placement*='top']::after {
        border-color: ${colors.white} transparent transparent transparent;
        border-width: 8px 8px 0 8px;
        top: -1px;
    }

    &[data-placement*='right'] {
        height: 1rem;
        left: 0;
        margin-left: -1rem;
        width: 1rem;
    }

    &[data-placement*='right']::before {
        border-color: transparent ${colors.grey400} transparent transparent;
        border-width: 8px 8px 8px 0;
        left: 5px;
    }

    &[data-placement*='right']::after {
        border-color: transparent ${colors.white} transparent transparent;
        border-width: 8px 8px 8px 0;
        left: 8px;
        top: 0;
    }

    &[data-placement*='left'] {
        height: 1rem;
        margin-right: -1rem;
        right: 0;
        width: 1rem;
    }

    &[data-placement*='left']::before {
        border-color: transparent transparent transparent ${colors.grey400};
        border-width: 8px 0 8px 8px;
        left: 2px;
    }

    &[data-placement*='left']::after {
        border-color: transparent transparent transparent ${colors.white};
        border-width: 8px 0 8px 8px;
        left: -1px;
        top: 0;
    }
`;
