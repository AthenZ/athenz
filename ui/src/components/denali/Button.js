/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { cx, css } from 'emotion';
import { rgba } from 'polished';
import { colors } from './styles/colors';
import { cssFontFamilies, cssFontWeights } from './styles/fonts';

// NOTE: Psuedo class priority:
// Link Visited Hover Focus Active
// "Lord Vader Hates Furry Animals"

const buttonBase = (props) => css`
    border: none;
    border-radius: 2px;
    box-shadow: 0 0 0 1px transparent inset, 0 0 0 0 ${colors.grey500} inset;
    box-sizing: border-box;
    cursor: pointer;
    display: inline-block;
    ${cssFontFamilies.default};
    ${cssFontWeights.normal};
    font-size: ${props.size === 'large'
        ? '18px'
        : props.size === 'regular'
        ? '14px'
        : '12px'};
    line-height: 1;
    margin: 5px;
    min-width: ${props.size === 'small' ? '70px' : undefined};
    outline: 0;
    padding: ${props.size === 'large'
        ? '12px 28px'
        : props.size === 'regular'
        ? '10px 24px'
        : '8px 18px'};
    text-align: center;
    text-shadow: none;
    text-transform: none;
    transition: ${!props.noanim ? 'all 0.2s ease-in' : undefined};
    vertical-align: baseline;
    white-space: nowrap;
    label: button;
    &:first-of-type {
        margin-left: 0;
    }
    &:disabled {
        background: ${props.dark
            ? rgba(colors.white, 0.1)
            : rgba(colors.grey800, 0.1)};
        color: ${props.dark
            ? rgba(colors.white, 0.2)
            : rgba(colors.grey800, 0.2)};
        cursor: not-allowed;
    }
`;

const buttonVariants = {
    primary: (props) => css`
        background: linear-gradient(
            to right,
            ${props.active ? colors.brand700 : colors.brand600},
            ${props.active ? colors.brand800 : colors.brand700}
        );
        color: ${colors.white};
        &:hover:not(:disabled) {
            background: linear-gradient(
                to right,
                ${colors.brand700},
                ${colors.brand800}
            );
        }
    `,
    secondary: (props) => css`
        background: ${props.dark
            ? props.active
                ? `linear-gradient(to right, ${rgba(
                      colors.brand700,
                      0.3
                  )}, ${rgba(colors.brand800, 0.3)})`
                : 'transparent'
            : props.active
            ? colors.brand200
            : 'transparent'};
        box-shadow: 0 0 0 1px ${colors.brand600} inset;
        color: ${props.dark ? colors.white : colors.brand600};
        &:disabled {
            box-shadow: ${props.dark
                ? `0 0 0 1px ${rgba(colors.white, 0.2)} inset`
                : `0 0 0 1px ${rgba(colors.grey800, 0.1)} inset`};
        }
        &:hover:not(:disabled) {
            background: ${props.dark
                ? `linear-gradient(to right, ${rgba(
                      colors.brand700,
                      0.3
                  )}, ${rgba(colors.brand800, 0.3)})`
                : colors.brand200};
        }
    `,
    danger: () => css`
        background: ${colors.red500};
        color: ${colors.white};
        &:hover:not(:disabled) {
            background: ${colors.red600};
        }
    `,
};

class Button extends React.PureComponent {
    render() {
        const {
            active,
            className,
            danger,
            dark,
            noanim,
            secondary,
            size,
            ...rest
        } = this.props;

        const cssProps = { active, dark, noanim, size };

        const makeCssButton = (props) => css`
            ${buttonBase(props)};
            ${(danger
                ? buttonVariants.danger
                : secondary
                ? buttonVariants.secondary
                : buttonVariants.primary)(props)};
        `;
        const cssButton = makeCssButton(cssProps);

        const classes = cx(cssButton, 'denali-button', className);

        return <button className={classes} {...rest} />;
    }
}

Button.propTypes = {
    /** Set active state */
    active: PropTypes.bool,
    /** Additonal class to apply to the button */
    className: PropTypes.string,
    /** Error color scheme */
    danger: PropTypes.bool,
    /** Dark color scheme (can be used in conjunction with 'secondary') */
    dark: PropTypes.bool,
    /** Disable the button (no clicks) */
    disabled: PropTypes.bool,
    /** Forward React ref to the `<input>` element */
    innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    /** Name of the `button` element */
    name: PropTypes.string,
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** Secondary color scheme */
    secondary: PropTypes.bool,
    /** Size of buttons (default is regular) */
    size: PropTypes.oneOf(['large', 'regular', 'small']),
    /** <button> type attribute */
    type: PropTypes.oneOf(['button', 'submit', 'reset']),
    /** <button> value attribute */
    value: PropTypes.any,
    /** Gets called when the button is clicked */
    onClick: PropTypes.func,
};

Button.defaultProps = {
    active: false,
    danger: false,
    dark: false,
    disabled: false,
    noanim: false,
    size: 'regular',
    secondary: false,
};

export default Button;
