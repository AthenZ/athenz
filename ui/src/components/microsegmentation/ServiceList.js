/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import styled from '@emotion/styled';
import DateUtils from '../utils/DateUtils';
import { withRouter } from 'next/router';
import Menu from '../denali/Menu/Menu';

const StyleTable = styled.table`
    width: 100%;
    text-align: center;
    border-spacing: 0;
    display: table;
    border-collapse: separate;
    border-color: black;
    box-sizing: border-box;
    margin-top: 5px;
    box-shadow: 0 1px 4px #d9d9d9;
    border: 5px solid #fff;
    -webkit-border-image: none;
    border-image: none;
    -webkit-border-image: initial;
    border-image: initial;
    max-height: 600px;
`;

const StyledDiv = styled.div`
    display: inline-block;
    overflow-y: scroll;
    max-height: 600px;
`;

const StyledTr = styled.tr`
    &:nth-child(even) {
        background-color: #3570f40d;
    }
`;

const StyledTh = styled.th`
    padding: 6px;
    border: 1px solid #dddddd;
`;

const StyledTd = styled.td`
    border: 1px solid #dddddd;
    padding: 5px;
    color: blue;
`;

const StyledLink = styled.a`
    cursor: pointer;
`;

const StyledInvalidLink = styled.a`
    color: red;
`;

const StyledMessageDiv = styled.div`
    min-width: 200px;
`;

const StyledMenu = styled(Menu)`
    min-width: 200px;
`;

class ServiceList extends React.Component {
    constructor(props) {
        super(props);
        this.api = props.api;
        this.localDate = new DateUtils();
        this.viewServiceDetails = this.viewServiceDetails.bind(this);
    }

    viewServiceDetails(item) {
        let index = item.lastIndexOf('.');
        let dom = item.substring(0, index);
        let service = item.substring(index + 1);
        this.props.router.push(
            `/domain/${dom}/service/${service}/instance/dynamic`,
            `/domain/${dom}/service/${service}/instance/dynamic`,
            { getInitialProps: true }
        );
    }

    render() {
        const { list } = this.props;
        let rows;
        rows = list.map((item, i) => {
            return (
                <StyledTr key={item + i + new Date().getTime()}>
                    <StyledTd>
                        {!item.includes('*') && (
                            <StyledLink
                                onClick={() => {
                                    this.viewServiceDetails(item);
                                }}
                            >
                                {' '}
                                {item}{' '}
                            </StyledLink>
                        )}

                        {item.includes('*') && (
                            <StyledMenu
                                placement='right'
                                trigger={
                                    <StyledInvalidLink>
                                        {' '}
                                        {item}{' '}
                                    </StyledInvalidLink>
                                }
                            >
                                <StyledMessageDiv>
                                    <p>
                                        Service is invalid as it contains
                                        wildcard characters.
                                    </p>
                                </StyledMessageDiv>
                            </StyledMenu>
                        )}
                    </StyledTd>
                </StyledTr>
            );
        });

        return (
            <StyledDiv
                key={'service-list'}
                data-testid={'segmentation-service-list'}
            >
                <StyleTable>
                    <StyledTr>
                        <StyledTh> Services </StyledTh>
                    </StyledTr>
                    {rows}
                </StyleTable>
            </StyledDiv>
        );
    }
}
export default withRouter(ServiceList);
