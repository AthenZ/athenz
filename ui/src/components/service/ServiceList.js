/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import styled from '@emotion/styled';
import { colors } from '../denali/styles';
import Button from '../denali/Button';
import NameUtils from '../utils/NameUtils';
import ServiceRow from './ServiceRow';
import Alert from '../denali/Alert';
import AddService from './AddService';
import DeleteModal from '../modal/DeleteModal';
import { MODAL_TIME_OUT } from '../constants/constants';
import RequestUtils from '../utils/RequestUtils';

const ServicesSectionDiv = styled.div`
    margin: 20px;
`;

const AddContainerDiv = styled.div`
    padding-bottom: 20px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-flow: row nowrap;
    float: right;
`;

const ServiceTable = styled.table`
    width: 100%;
    border-spacing: 0;
    display: table;
    border-collapse: separate;
    border-color: ${colors.grey600};
`;

const TableHeadStyled = styled.th`
    text-align: ${(props) => props.align};
    border-bottom: 2px solid ${colors.grey500};
    color: ${colors.grey600};
    font-weight: 600;
    font-size: 0.8rem;
    padding-bottom: 5px;
    vertical-align: top;
    text-transform: uppercase;
    padding: 5px 0 5px 15px;
    word-break: break-all;
`;

const StyledAnchor = styled.a`
    color: #3570f4;
    text-decoration: none;
    cursor: pointer;
`;

export default class ServiceList extends React.Component {
    constructor(props) {
        super(props);
        this.api = props.api;
        this.onCancelDeleteService = this.onCancelDeleteService.bind(this);
        this.toggleAddService = this.toggleAddService.bind(this);
        this.reloadServices = this.reloadServices.bind(this);
        this.closeModal = this.closeModal.bind(this);
        this.onSubmitDeleteService = this.onSubmitDeleteService.bind(this);
        this.state = {
            list: props.services || [],
        };
    }

    onSubmitDeleteService() {
        this.api
            .deleteService(
                this.props.domain,
                this.state.deleteServiceName,
                this.props._csrf
            )
            .then(() => {
                this.reloadServices(
                    `Successfully deleted service ${this.state.deleteServiceName}`,
                    true
                );
            })
            .catch((err) => {
                this.setState({
                    errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                });
            });
    }

    onCancelDeleteService() {
        this.setState({
            showDelete: false,
            deleteServiceName: null,
        });
    }

    onClickDeleteService(serviceName) {
        this.setState({
            showDelete: true,
            deleteServiceName: serviceName,
            errorMessage: null,
        });
    }
    //successMessage is only name of new service when adding a service
    reloadServices(successMessage, showSuccess) {
        this.api
            .getServices(this.props.domain)
            .then((data) => {
                this.setState({
                    list: data,
                    showAddService: false,
                    showSuccess,
                    successMessage,
                    showDelete: false,
                });
                setTimeout(
                    () =>
                        this.setState({
                            showSuccess: false,
                        }),
                    MODAL_TIME_OUT
                );
            })
            .catch((err) => {
                this.setState({
                    errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                });
            });
    }

    closeModal() {
        this.setState({ showSuccess: false });
    }

    toggleAddService() {
        this.setState({
            showAddService: !this.state.showAddService,
        });
    }

    render() {
        const { domain } = this.props;
        const left = 'left';
        const center = 'center';
        const rows = this.state.list.map((item, i) => {
            const serviceName = NameUtils.getShortName('.', item.name);
            let newService = serviceName === this.state.successMessage;
            let onClickDeleteService = this.onClickDeleteService.bind(
                this,
                serviceName
            );
            let color = '';
            if (i % 2 === 0) {
                color = colors.row;
            }
            let toReturn = [];
            toReturn.push(
                <ServiceRow
                    serviceName={serviceName}
                    domainName={domain}
                    modified={item.modified}
                    newService={newService}
                    color={color}
                    api={this.api}
                    key={item.name}
                    _csrf={this.props._csrf}
                    onClickDeleteService={onClickDeleteService}
                    featureFlag={this.props.featureFlag}
                />
            );
            return toReturn;
        });
        let addService = this.state.showAddService ? (
            <AddService
                showAddService={this.state.showAddService}
                onCancel={this.toggleAddService}
                onSubmit={this.reloadServices}
                domain={domain}
                api={this.api}
                _csrf={this.props._csrf}
                pageConfig={this.props.pageConfig}
            />
        ) : (
            ''
        );
        return (
            <ServicesSectionDiv data-testid='service-list'>
                <AddContainerDiv>
                    <div>
                        <Button secondary onClick={this.toggleAddService}>
                            Add Service
                        </Button>
                        {addService}
                    </div>
                </AddContainerDiv>
                <ServiceTable>
                    <thead>
                        <tr>
                            <TableHeadStyled align={left}>
                                Service
                            </TableHeadStyled>
                            <TableHeadStyled align={left}>
                                Modified Date
                            </TableHeadStyled>
                            <TableHeadStyled align={center}>
                                Instances
                            </TableHeadStyled>
                            <TableHeadStyled align={center}>
                                Public Keys
                            </TableHeadStyled>
                            <TableHeadStyled align={center}>
                                Providers
                            </TableHeadStyled>
                            <TableHeadStyled align={center}>
                                Delete
                            </TableHeadStyled>
                        </tr>
                    </thead>
                    <tbody>{rows}</tbody>
                </ServiceTable>
                {this.state.showSuccess ? (
                    <Alert
                        isOpen={this.state.showSuccess}
                        title={this.state.successMessage}
                        onClose={this.closeModal}
                        type='success'
                    />
                ) : null}
                {this.state.showDelete ? (
                    <DeleteModal
                        name={this.state.deleteServiceName}
                        isOpen={this.state.showDelete}
                        cancel={this.onCancelDeleteService}
                        submit={this.onSubmitDeleteService}
                        errorMessage={this.state.errorMessage}
                        message={
                            'Are you sure you want to permanently delete the Service '
                        }
                    />
                ) : null}
            </ServicesSectionDiv>
        );
    }
}
