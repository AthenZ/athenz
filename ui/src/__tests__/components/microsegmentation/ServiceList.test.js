/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import { colors } from '../../../components/denali/styles';
import ServiceList from "../../../components/microsegmentation/ServiceList";
import API from '../../../api';

describe('ServiceList', () => {
    it('should render', () => {
        let details = [
                "serviceB",
                "serviceC",
                "serviceD"
            ];

        let domain = 'domain';
        const { getByTestId } = render(
            <table>
                <tbody>
                   <ServiceList
                       list={details}
                       api={API()}
                       domain={domain}
                   />
                </tbody>
            </table>
        );
        const ruleRow = getByTestId('segmentation-service-list');

        expect(ruleRow).toMatchSnapshot();
    });
});
