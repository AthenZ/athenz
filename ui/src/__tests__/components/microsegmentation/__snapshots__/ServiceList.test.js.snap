// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceList should render 1`] = `
.emotion-0 {
  display: inline-block;
  overflow-y: scroll;
  max-height: 600px;
}

.emotion-2 {
  width: 100%;
  text-align: center;
  border-spacing: 0;
  display: table;
  border-collapse: separate;
  border-color: black;
  box-sizing: border-box;
  margin-top: 5px;
  box-shadow: 0 1px 4px #d9d9d9;
  border: 5px solid #fff;
  -webkit-border-image: none;
  border-image: none;
  -webkit-border-image: initial;
  border-image: initial;
  max-height: 600px;
}

.emotion-4:nth-child(even) {
  background-color: #3570f40d;
}

.emotion-6 {
  padding: 6px;
  border: 1px solid #dddddd;
}

.emotion-10 {
  border: 1px solid #dddddd;
  padding: 5px;
  color: blue;
}

.emotion-12 {
  cursor: pointer;
}

<div
  class="emotion-0 emotion-1"
  data-testid="segmentation-service-list"
>
  <table
    class="emotion-2 emotion-3"
  >
    <tr
      class="emotion-4 emotion-5"
    >
      <th
        class="emotion-6 emotion-7"
      >
         Services 
      </th>
    </tr>
    <tr
      class="emotion-4 emotion-5"
    >
      <td
        class="emotion-10 emotion-11"
      >
        <a
          class="emotion-12 emotion-13"
        >
           
          serviceB
           
        </a>
      </td>
    </tr>
    <tr
      class="emotion-4 emotion-5"
    >
      <td
        class="emotion-10 emotion-11"
      >
        <a
          class="emotion-12 emotion-13"
        >
           
          serviceC
           
        </a>
      </td>
    </tr>
    <tr
      class="emotion-4 emotion-5"
    >
      <td
        class="emotion-10 emotion-11"
      >
        <a
          class="emotion-12 emotion-13"
        >
           
          serviceD
           
        </a>
      </td>
    </tr>
  </table>
</div>
`;
